### PR TITLE
[Go] Implement numbers floating point concept exercise

### PR DESCRIPTION
**Pull request by imanc**
Using C# exercise as base, I replicate this in Go.

### CODE REVIEW

File: languages/go/exercises/concept/numbers-floating-point/numbers-floating-point_test.go
Comment (imanc) [inline languages/go/exercises/concept/numbers-floating-point/numbers-floating-point_test.go:28]: This test case from C# fails in Go with a float64

File: languages/go/exercises/concept/numbers-floating-point/numbers-floating-point_test.go
Comment (bitfield) [inline languages/go/exercises/concept/numbers-floating-point/numbers-floating-point_test.go:14]: We can use spaces in these names, which would make them a bit easier to read, wouldn't it?

**Review by bitfield — CHANGES_REQUESTED** [review]
This is a great start! However, I think using floating-point for currency calculations is problematic, and a classic source of bugs. It would be great to start brand-new programmers out on the right path by avoiding that, wouldn't it?

File: languages/go/exercises/concept/numbers-floating-point/.docs/instructions.md
Comment (imanc) [reply languages/go/exercises/concept/numbers-floating-point/.docs/instructions.md:26]: Hi, yes I agree but I think I completely miss the point: I thought I need to copy the exact exercise transposing to golang. Can we change completely instead?

File: languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md
Comment (tehsphinx) [reply languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md:7]: I disagree as this is not an intro to numbers, but to floating point numbers. It needs to at least mention `float32`.

File: languages/go/exercises/concept/numbers-floating-point/.meta/design.md
Comment (tehsphinx) [reply languages/go/exercises/concept/numbers-floating-point/.meta/design.md:5]: Agree, although this is how it is formulated in the issue templates for new concepts (which I change every time I create an issue) and probably also in the C#.

File: languages/go/exercises/concept/numbers-floating-point/.docs/instructions.md
Comment (tehsphinx) [reply languages/go/exercises/concept/numbers-floating-point/.docs/instructions.md:26]: We have 2 options here I think:
1) deviate from the original C# exercise in the story we tell here so it is not about currency
2) tell the student that normally we wouldn't use floats for representing currency but here we will still do it for the sake of this exercise

I would vote for option 1) and it looks to me we all agree on that.

@imanc: can you think of another storyline here?

File: languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md
Comment (tehsphinx) [reply languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md:1]: I think this is from the original C# exercise, so we should also change it there if you think this is a problem...

File: languages/go/exercises/concept/numbers-floating-point/.meta/example.go
Comment (tehsphinx) [reply languages/go/exercises/concept/numbers-floating-point/.meta/example.go:5]: It is not an example the user will ever see, but I agree it would be nice to have good comments anyway.

File: languages/go/exercises/concept/numbers-floating-point/numbers-floating-point.go
Comment (tehsphinx) [inline languages/go/exercises/concept/numbers-floating-point/numbers-floating-point.go:3]: I would vote to have a space after the two slashes...

File: languages/go/exercises/concept/numbers-floating-point/numbers-floating-point_test.go
Comment (tehsphinx) [reply languages/go/exercises/concept/numbers-floating-point/numbers-floating-point_test.go:28]: floats can not be tested with `==` or `!=`. They need to be tested as:

```go
eps := 0.01
if eps < math.Abs(got - tt.want) {
    t.Errorf(...)
}
```

Comment (tehsphinx) [reply languages/go/exercises/concept/numbers-floating-point/numbers-floating-point_test.go:28]: where eps (epsilon) is the acceptable tolerance

File: languages/go/exercises/concept/numbers-floating-point/.docs/instructions.md
Comment (bitfield) [reply languages/go/exercises/concept/numbers-floating-point/.docs/instructions.md:26]: What were the C# people thinking in the first place? 🤔

File: languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md
Comment (bitfield) [reply languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md:1]: Right, but that's not part of this PR. I can only review the stuff I see.

File: languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md
Comment (bitfield) [reply languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md:7]: Well, if `float32` is never used in Go, I'm concerned that including it in this exercise would actually make the exercise _less_ valuable. They say the secret of being a bore is tell _everything_; that's also the secret of being a bad teacher. At this early stage I think it's wise to keep the number of new concepts to a minimum, and only introduce those things which bring great value for the student in writing real programs.

File: languages/go/exercises/concept/numbers-floating-point/.meta/example.go
Comment (bitfield) [reply languages/go/exercises/concept/numbers-floating-point/.meta/example.go:5]: What's it for, then?

Comment (tehsphinx) [reply languages/go/exercises/concept/numbers-floating-point/.meta/example.go:5]: For reviews and discussions it is what I look at first to check what the exercise ist about. It is far easier to get a quick grasp than to read the intro and think of a solution myself. Especially when it comes to see what other concepts are needed, when writing an analyzer for it, etc.
Actually to me as a maintainer it is the most important document of a concept exercise.

Comment (tehsphinx) [reply languages/go/exercises/concept/numbers-floating-point/.meta/example.go:5]: But yes, it's not for the student -- at least for now. I don't know of any plans to show this to students.

File: languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md
Comment (tehsphinx) [reply languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md:7]: I don't see the harm in mentioning that it exists.

Note: I use float32 in my services as I have IOT devices that transmit float32. Also I have to build one of my applications for windows 32bit where float64 works but if its not necessary, why use it? So... I'm actually using float32 quite a bit. I'm not dismissing that the standard library expects float64 everywhere and that float32 is definitely not widespread. Still I think the information that a float32 exists doesn't hurt.

File: languages/go/exercises/concept/numbers-floating-point/.docs/instructions.md
Comment (tehsphinx) [reply languages/go/exercises/concept/numbers-floating-point/.docs/instructions.md:26]: doesn't C# have a decimal type?

File: languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md
Comment (bitfield) [reply languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md:7]: Fair enough! Glad to hear there _is_ a use for it!

**Comment by tehsphinx** [comment]
@ErikSchierboom: what's the best way to continue here?

If I merge this (provided @bitfield unblocks it), we will probably have some checks fail on master again. If I just create a new branch and recommit from scratch, @imanc is removed from the git history... But I can't fix it in his branch or rebase, etc...

**Comment by kotp** [comment]
> If I merge this (provided @bitfield unblocks it), we will probably have some checks fail on master again. If I just create a new branch and recommit from scratch, @imanc is removed from the git history... But I can't fix it in his branch or rebase, etc...

You could cheery-pick the commits you want that come from @imanc and the attribution would still be in the history.  You could also base your new branch on the work done here, and the history would stay, and so attribution would survive.

File: languages/go/exercises/concept/numbers-floating-point/.meta/example.go
Comment (ErikSchierboom) [reply languages/go/exercises/concept/numbers-floating-point/.meta/example.go:5]: I've created a [PR to help clarify this a bit](https://github.com/exercism/v3/pull/1567).

File: languages/go/exercises/concept/numbers-floating-point/.meta/design.md
Comment (ErikSchierboom) [reply languages/go/exercises/concept/numbers-floating-point/.meta/design.md:5]: Well this is a bit semantics, but you _could_ look at it as there being a general Concept (floating-point numbers) and then a language-specific implementation. That does not mean you have to understand _how_ it is implemented, but you should indeed know how to use it (which means knowing though what constructs/types the concept is implemented in the language). Feel free to reword though.

File: languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md
Comment (ErikSchierboom) [reply languages/go/exercises/concept/numbers-floating-point/.docs/introduction.md:7]: > Note: I use float32 in my services as I have IOT and other hardware devices that transmit float32. Also I have to build one of my applications for windows 32bit where float64 works but if its not necessary, why use it? So... I'm actually using float32 quite a bit. I'm not dismissing that the standard library expects float64 everywhere and that float32 is definitely not widespread. Still I think the information that a float32 exists doesn't hurt.

It might be useful to add this information to the `after.md` document!

File: languages/go/exercises/concept/numbers-floating-point/.docs/instructions.md
Comment (ErikSchierboom) [reply languages/go/exercises/concept/numbers-floating-point/.docs/instructions.md:26]: > What were the C# people thinking in the first place?

Good question :) In C#, there are three floating-point types: `double`/`float`/`decimal`. The latter is exclusively used in financial code, and is the recommend type for doing so. One should never use `double`/`float` for storing money though, but that is also explained in the exercise. In short, while the exercise might work well for C#/F#, I can totally see why it wouldn't work here. As a consequence, it makese sense to adapt/rewrite the theme to have something that works for Go. Note that this can and will happen when porting exercises, as languages will implement concepts differently.

**Review by ErikSchierboom — COMMENTED** [review]
Thanks a lot for working on this @imanc! It appears that there may be some issues with porting the C#/F# exercise, but they are totally solvable. I think you and the PR reviewers can probably figure out how to make this an exercise that works well for Go.

**Comment by ErikSchierboom** [comment]
@imanc Is there anything I can do to help with this PR?

**Comment by ErikSchierboom** [comment]
This Friday (Jan 29th 2021), we're going to explode this repo back into the individual track repositories and archive this repo.

As such, this PR needs to be merged by Friday 9am UTC, or it will need to be manually reopened against the track repo.

**Pull request closed without merging**
